mypy/subtypes.py

```diff
@@ -143,10 +143,15 @@ def visit_instance(self, left: Instance) -> bool:
             rname = right.type.fullname()
             if not left.type.has_base(rname) and rname != 'builtins.object':
                 return False
-
             # Map left type to corresponding right instances.
             t = map_instance_to_supertype(left, right.type)
-
+            # Check that one of the types does not have type args or
+            # the number of type args is the same and
+            # each left type arg is acceptable in place of the matching right one
+            if not t.args or not right.args:
+                return True
+            if len(t.args) != len(right.args):
+                return False
```

Comment:
Instances that cause problems represent fallback for various tuples. I tried to force all tuples to be created with `len(fallback) == len(tuple.items())` by changing stuff like

    # checker.py
    return TupleType(type_parameters, self.named_type('builtins.tuple'))` 

into 

    return TupleType(
        type_parameters,
       self.named_generic_type('builtins.tuple', [AnyType()]*len(type_parameters))

But (a) I'm not sure it's even correct; and (b) star args need to have variadic fallback.

So I could not get rid of assertion failure.

I pushed the code I tried to a [separate branch](https://github.com/pkch/mypy/tree/tuple-length), if you like to take a look.